pkg/flowaggregator/s3uploader/s3uploader.go

```diff
@@ -0,0 +1,464 @@
+// Copyright 2022 Antrea Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package s3uploader
+
+import (
+	"bytes"
+	"compress/gzip"
+	"context"
+	"fmt"
+	"io"
+	"math/rand"
+	"sync"
+	"time"
+
+	"github.com/aws/aws-sdk-go-v2/aws"
+	awsconfig "github.com/aws/aws-sdk-go-v2/config"
+	s3manager "github.com/aws/aws-sdk-go-v2/feature/s3/manager"
+	"github.com/aws/aws-sdk-go-v2/service/s3"
+	ipfixentities "github.com/vmware/go-ipfix/pkg/entities"
+	"k8s.io/klog/v2"
+
+	config "antrea.io/antrea/pkg/config/flowaggregator"
+	"antrea.io/antrea/pkg/flowaggregator/flowrecord"
+)
+
+const (
+	bufferFlushTimeout = 1 * time.Minute
+	bufferQueueSize    = 3
+	seed               = 1
+)
+
+var (
+	nameRand *rand.Rand
+)
+
+type stopPayload struct {
+	flushQueue bool
+}
+
+type S3UploadProcess struct {
+	bucketName       string
+	bucketPrefix     string
+	region           string
+	compress         *bool
+	maxRecordPerFile int32
+	// uploadInterval is the interval between batch uploads
+	uploadInterval time.Duration
+	// uploadTicker is a ticker, containing a channel used to trigger batchUploadAll() for every uploadInterval period
+	uploadTicker *time.Ticker
+	// stopCh is the channel to receive stop message
+	stopCh chan stopPayload
+	// exportWg is to ensure that all messages have been flushed from the queue when we stop
+	exportWg             sync.WaitGroup
+	exportProcessRunning bool
+	// mutex protects configuration state from concurrent access
+	mutex sync.Mutex
+	// bufferMutex is for concurrency between writing and reading records from currentBuffer
+	bufferMutex sync.Mutex
```

Comment:
the purpose of `bufferMutex` is not clear to me. We should never have concurrent read and writes to the buffer. We write to it until it is full or a certain amount of time has elapsed, then we add it to the `bufferQueue`. After that, we should not write to it again.